File: 剑指offer/每日一题/leetcode2383.py
from typing import List


class Solution:
    def minNumberOfHours(self, initialEnergy: int, initialExperience: int, energy: List[int],
                         experience: List[int]) -> int:
        sm = sum(energy)
        trainingHours = 0 if initialEnergy > sm else sm + 1 - initialEnergy
        for e in experience:
            if initialExperience <= e:
                trainingHours += 1 + (e - initialExperience)
                initialExperience = 2 * e + 1
            else:
                initialExperience += e
        return trainingHours


def main():
    initialEnergy = 7
    initialExperience = 1
    energy = [4, 3, 1]
    experience = [2, 1, 3]

    solution = Solution()
    min_hours = solution.minNumberOfHours(initialEnergy, initialExperience, energy, experience)
    print(f"Minimum number of training hours needed: {min_hours}")


if __name__ == '__main__':
    main()

File: 剑指offer/每日一题/leetcode2383.cpp

//
// Created by StarkLu on 2023/3/13.
//
#include <vector>
using namespace std;
class Solution {
public:
    int minNumberOfHours(int initialEnergy, int initialExperience, vector<int> &energy, vector<int> &experience) {
        int sum = 0;
        for(auto e:energy){
            sum += e;
        }
        int time  = sum <= initialEnergy ? 0: (initialEnergy - sum) + 1;
        for(auto ex:experience){
            if(initialExperience <= ex){
                time += 1 + (ex - initialExperience);
                initialExperience = 2 * ex + 1;
            }
            else{
                initialExperience += ex;
            }
        }
        return time;
    }
};